# Exercise 11

## Exercise 11 (no programming)

Make the frequency analysis of the text by normalizing all the words in the text.

[How to make the task.](https://courses.openedu.ru/assets/courseware/v1/84bcb7c4563a6772addcb93e341e0f05/asset-v1:ITMOUniversity+INTROMLAAISBENG+spring_2023_ITMO_mag+type@asset+block/Task_1_not_programmers__EN_instruction_.pdf)

Click [**the link**](https://courses.openedu.ru/assets/courseware/v1/b4c7823727d4aeb75d1e0b5c5355ae85/asset-v1:ITMOUniversity+INTROMLAAISBENG+spring_2023_ITMO_mag+type@asset+block/Task_1_EN_not_progr_leskov.txt) to see a piece of the Russian classical literature. Your task is to normalize words using [**the tool**](http://textanalysisonline.com/spacy-word-lemmatize) and find the frequencies of the specified words using [**the service**](https://planetcalc.com/3205/). Enter only the frequency of the word itself (neglect the words with attached symbols).

**Enter the frequency of the word "molva"**
- ✅ 6

 
**Enter the frequency of the word "russia"**
- ✅ 7

**Enter the frequency of the word "carriage"**
- ✅ 6
 
**Enter the frequency of the word "flea"**
- ✅ 33 
 

## Exercise 11 (programming)

Click the link to see the task on adding the code parts:

Notebook in *.pynb format with some examples of commands to make the exercise.
We recommend to use Google colab  to work with Notebook. 

Your task is to add some parts of code to the example considered. The right answers are in the Notebook. Please proceed to the individual task only when you obtain a correctly functioning code.

## Demo Part

> For Individual task scroll below

**Data URL**

In [1]:
DATA_URL = "http://www.gutenberg.org/files/913/913-0.txt"

In [2]:
!pip install -q nltk # ==3.2.5


[notice] A new release of pip is available: 23.0.1 -> 23.1.2
[notice] To update, run: pip install --upgrade pip


In [ ]:
# !cd ../
# !unzip /usr/local/lib/python3.10/dist-packages/nltk_data/corpora/wordnet.zip -d ../usr/local/lib/python3.10/dist-packages/nltk_data/corpora/nltk_data/corpora/

unzip:  cannot find or open ../usr/local/lib/python3.10/dist-packages/nltk_data/corpora/wordnet.zip, ../usr/local/lib/python3.10/dist-packages/nltk_data/corpora/wordnet.zip.zip or ../usr/local/lib/python3.10/dist-packages/nltk_data/corpora/wordnet.zip.ZIP.


Loading *NLTK*'s 'wordnet'

In [3]:
import nltk
from nltk.corpus import wordnet
nltk.download("wordnet")

lemmatizer = nltk.stem.wordnet.WordNetLemmatizer()
lemmatizer

[nltk_data] Downloading package wordnet to /home/datalore/nltk_data...


<WordNetLemmatizer>

Downloading the text for the task via `urllib`

In [4]:
# import urllib.request
from urllib.request import build_opener, HTTPCookieProcessor, Request


# opener = urllib.request.URLopener({})

opener = build_opener(HTTPCookieProcessor())
request = Request(DATA_URL)
resource = opener.open(request, timeout=30)
# content = response.read()


# resource = opener.open(DATA_URL)
charset = resource.headers.get_content_charset()
print("Charset", charset)
raw_text = resource.read()

# raw_text = urllib.urlopen(DATA_URL).read()

if charset:
  raw_text = raw_text.decode(resource.headers.get_content_charset())
else:
  raw_text = raw_text.decode("utf-8")

raw_text[:100]

Charset None


'\ufeffThe Project Gutenberg EBook of A Hero of Our Time, by M. Y. Lermontov\r\n\r\nThis eBook is for the use '

In [15]:
import requests

response = requests.get(DATA_URL)
raw_text = response.text
raw_text[:100]

'The Project Gutenberg EBook of A Hero of Our Time, by M. Y. Lermontov\r\n\r\nThis eBook is for the us'

Removing the book ending (Gutenberg legal information)

In [5]:
import re

clean_pattern = re.compile("End of the Project Gutenberg EBook.*")
cleaner_text =  re.sub(clean_pattern, "", raw_text.replace("\n", " ").replace("\r", " "))
cleaner_text[-100:]

'ases, equivalent to: “Men are fools, fortune is  blind, and life is not worth a straw.”]            '

Splitting the text into tokens with a little help from [NLTK](https://nltk.org/).

In [6]:
from nltk.tokenize import word_tokenize
import nltk
nltk.download('punkt')

tokens =  word_tokenize(cleaner_text)
"A total of %d 'tokens'" % len(tokens)

[nltk_data] Downloading package punkt to /home/datalore/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


"A total of 72566 'tokens'"

Now we are about to **lemmatize the tokens**. Please note that for better results we should have first PoS-tagged the text (e.g. with NLTK as well, [please refer to the book and the docs](https://www.nltk.org/book/ch05.html)). `WordNetLemmatizer` would work best with PoS tags provided. However, to make things short and simple, we won't do it as of now.



---


## Task \#1
using Python's standard library's `str.isalpha` modify the code below to remove all non-letter tokens from sentences.

---


In [35]:
"a1aa".isalpha()

False

In [7]:
from tqdm.notebook import tqdm

lemmas = [lemmatizer.lemmatize(lemma) for lemma in tqdm(tokens) if lemma.isalpha()]
lemmas[-10:]

['fortune', 'is', 'blind', 'and', 'life', 'is', 'not', 'worth', 'a', 'straw']

Let us check if everything is correct. Total corpus size should be equal to 59087

In [8]:
len(lemmas)

59087

To continue working on this assignment make sure those numbers match!


---

## Task #2

Using `lemmas`, english NLTK stopwords (`nltk.corpus.stopwords`) and `nltk.FreqDist`, compute **the fraction of the stopwords** in the top-100 most frequent words in the text. 

E.g. if there were just **3 stopwords in 100** most frequent words in the text, the answer would be **0.03**

**Googling how to work with NLTK is encouraged.**

---


In [15]:
import nltk
from nltk import FreqDist
from nltk.corpus import stopwords
nltk.download("stopwords")
STOPWORDS = set(stopwords.words("english"))
stopwords.words("english")[:5]

[nltk_data] Downloading package stopwords to
[nltk_data]     /home/datalore/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


['i', 'me', 'my', 'myself', 'we']

Check yourself: stopwords rate in 100 most frequent words in the text should be 0.66

In [16]:
from nltk.probability import FreqDist

fdist = FreqDist(lemma for lemma in lemmas)

In [12]:
len(fdist)

5971

In [17]:
cnt = 0
for i, val in enumerate(fdist):
    if i + 1 == 100:
        break
    if val in STOPWORDS:
        cnt+=1

print(f"{cnt/100}")

0.66




---


## Task #3
Compute how many tokens occur in the text **strictly greater than** 50 times.

---


In [26]:
mapka = dict()

for token in lemmas:
    if token not in mapka:
        mapka[token] = 0
    mapka[token] += 1

In [27]:
cnt_gr_50 = 0

for val in mapka.values():
    if val > 50:
        cnt_gr_50 += 1

print(f"{cnt_gr_50}")

149


Chek yourself: 149

## Individual task

Click the link to see an example of the Russian classical literature. Your task is to analyze the text.

### Helpers

In [28]:
EX_DATA_URL = "http://www.gutenberg.org/files/1081/1081-h/1081-h.htm"

from urllib.request import build_opener, HTTPCookieProcessor, Request

opener = build_opener(HTTPCookieProcessor())
request = Request(EX_DATA_URL)
resource = opener.open(request, timeout=30)

charset = resource.headers.get_content_charset()
print("Charset", charset)
raw_ex_text = resource.read()

if charset:
  raw_ex_text = raw_ex_text.decode(resource.headers.get_content_charset())
else:
  raw_ex_text = raw_ex_text.decode("utf-8")

raw_ex_text[:100]

Charset None


'<?xml version="1.0" encoding="utf-8"?>\r\n\r\n<!DOCTYPE html\r\n   PUBLIC "-//W3C//DTD XHTML 1.0 Strict//E'

In [29]:
import re

clean_pattern = re.compile("End of the Project Gutenberg EBook.*")
cleaner_ex_text =  re.sub(clean_pattern, "", raw_ex_text.replace("\n", " ").replace("\r", " "))
cleaner_ex_text[-100:]

'districts.]      </p>      <p>        <br /><br />      </p>  <pre xml:space="preserve">            '

In [31]:
from nltk.tokenize import word_tokenize
import nltk
nltk.download('punkt')

ex_tokens = word_tokenize(cleaner_ex_text)
"A total of %d 'tokens'" % len(ex_tokens)

[nltk_data] Downloading package punkt to /home/datalore/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


"A total of 202226 'tokens'"

## Task 1 - Enter the number of letter tokens

In [32]:
ex_lemmas = [lemmatizer.lemmatize(lemma) for lemma in tqdm(ex_tokens) if lemma.isalpha()]
len(ex_lemmas)

151419

## Task 2 - What is the share of stop-words of 200 most frequent tokens in the text?

In [33]:
ex_fdist = FreqDist(lemma for lemma in ex_lemmas)

In [34]:
ex_cnt = 0
for i, val in enumerate(ex_fdist):
    if i + 1 == 200:
        break
    if val in STOPWORDS:
        ex_cnt += 1

print(f"{ex_cnt/200}")

0.45


## Task 3 - What is the amount of tokens encountered in the text strictly greater than 20 times?

In [36]:
ex_mapka = dict()

for token in ex_lemmas:
    if token not in ex_mapka:
        ex_mapka[token] = 0
    ex_mapka[token] += 1

In [38]:
cnt_gr_20 = 0

for val in ex_mapka.values():
    if val > 20:
        cnt_gr_20 += 1

print(f"{cnt_gr_20}")

791
